presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java

```diff
@@ -552,6 +558,13 @@ public boolean transitionToFinishing()
             return false;
         }
 
+        try {
+            metadata.endQuery(session);
```

Comment:
The `endQuery` failure might mean that some of the resources weren't cleared. The query should fail in this case. If the query is failed there is nothing to commit. We need to do `rollback` instead.